### PR TITLE
ramips: Add support for Ruijie RG-EW1200G PRO

### PR DESCRIPTION
**Pull request by JakauBahuslau**
ramips: add support for Ruijie RG-EW1200G PRO

- SoC:      MediaTek MT7621
- Flash:    16MB (Macronix MX25L12805D)
- RAM:     128MB
- Serial:   As marked on PCB, baudrate is 57600, DO NOT CONNECT 3.3V!!!
- Ethernet: 4x 10/100/1000 Mbps (3x LAN + WAN)
- WIFI0:    MT7615 2.4GHz 802.11b/g/n
- WIFI1:    MT7615 5GHz 802.11ac
- Antennas: 6x external (3 per radio), non-detachable
- LEDs:     Programmable power-LED (blue-colored)
- Buttons:  Reset

INSTALLATION:

1. Get rootshell using insructions from https://gist.github.com/ZIKH26/18693c67ee7d2f8d2c60231b19194c37

2. Download and flash image
On computer:
python -m http.server
On router:
cd /tmp
wget http://<IP>:8000/factory.bin
mtd -r write factory.bin firmware

Device should reboot at this point.

Reverting to stock:
Download archive with firmware from Ruijie's site and
get .bin file from it. Then write that binary to firmware
partition. After reboot, factory-reset the router using 
reset button.

Signed-off-by: Yahor Leonenka staryjakau@gmail.com

### CODE REVIEW

**Comment by JakauBahuslau** [comment]
@musashino205 recreated pull request

**Comment by musashino205** [comment]
Please don't create a new PR instead of updating the old one... Existing PR can be updated by ```git push -f```.

https://openwrt.org/docs/guide-developer/frequent-pr-mistakes#do_not_multi-post_or_re-post_prs

And some guides for working on GitHub:
https://openwrt.org/docs/guide-developer/working-with-github-pr

**Comment by JakauBahuslau** [comment]
> Please don't create a new PR instead of updating the old one

Unfortunately, I already created a new one. Sorry.

**Comment by JakauBahuslau** [comment]
Tried to reopen old PR from your guide, unfortunately, nothing happened. 

In case of what, here's a link to old PR: https://github.com/openwrt/openwrt/pull/14746

**Comment by JakauBahuslau** [comment]
Fixed 2 errors. The first one was syntax error in my commit message, it said that it needed my real name, not my nickname. The second one was syntax error in DTS. While fixing what @musashino205 said I needed to do, I forgot to add a curly bracket in one single place.

**Comment by musashino205** [comment]
Please copy the content in the first PR comment to the commit message. Specification, install instruction, and others.

**Comment by JakauBahuslau** [comment]
> Please copy the content in the first PR comment to the commit message. Specification, install instruction, and others.

Done

**Comment by JakauBahuslau** [comment]
@musashino205 could you please rerun all these tests? There was an error that was being caused because my username was also set as my name. I changed it and now the test should pass (or I hope so)

**Comment by musashino205** [comment]
I'm not a team member of OpenWrt project and I don't have a permission to do it.
BTW, did you read the error message of tests? 

```
Signed-off-by is missing or doesn't match author (should be 'Signed-off-by: JakaŭBahusłaŭ <117989861+JakauBahuslau@users.noreply.github.com>')
```

**Comment by JakauBahuslau** [comment]
> I'm not a team member of OpenWrt project and I don't have a permission to do it. BTW, did you read the error message of tests?
Of cource I did. The one you pointed to is actually stand the second.The first one is: `
Author name (JakaŭBahusłaŭ) need to be your real name 'firstname lastname'` 
After that, I changed my 'Name' section toy real name.

File: target/linux/ramips/dts/mt7621_ruijie_rg-ew1200g-pro.dts [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+
+#include <dt-bindings/gpio/gpio.h>
+#include <dt-bindings/input/input.h>
+#include <dt-bindings/leds/common.h>
+
+/ {
+	compatible = "ruijie,rg-ew1200g-pro", "mediatek,mt7621-soc";
+	model = "Ruijie RG-EW1200G PRO";
+
+	leds {
```
Comment (DragonBluep) [inline target/linux/ramips/dts/mt7621_ruijie_rg-ew1200g-pro.dts:13]: Please sort these nodes alphabetically.

File: target/linux/ramips/dts/mt7621_ruijie_rg-ew1200g-pro.dts [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+};
+
+&spi0 {
+	status = "okay";
+
+	flash@0 {
+		compatible = "jedec,spi-nor";
+		reg = <0>;
+		#address-cells = <1>;
+		#size-cells = <1>;
```
Comment (DragonBluep) [inline target/linux/ramips/dts/mt7621_ruijie_rg-ew1200g-pro.dts:52]: `#address-cells` and `#size-cells` should be removed as `partitions` node doesn't have `reg` property.

File: target/linux/ramips/dts/mt7621_ruijie_rg-ew1200g-pro.dts [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+				read-only;
+			};
+
+			partition@50000 {
+				label = "u-boot-env";
+				reg = <0x50000 0x10000>;
+				read-only;
+			};
+
+			factory: partition@60000 {
```
Comment (DragonBluep) [inline target/linux/ramips/dts/mt7621_ruijie_rg-ew1200g-pro.dts:70]: Drop `factory: ` label.

File: target/linux/ramips/dts/mt7621_ruijie_rg-ew1200g-pro.dts [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+				reg = <0x60000 0x10000>;
+				read-only;
+
+				nvmem-layout {
+					compatible = "fixed-layout";
+					#address-cells = <1>;
+					#size-cells = <1>;
+
+					eeprom_factory_0: eeprom@0 {
+						reg = <0x0 0x400>;
```
Comment (DragonBluep) [inline target/linux/ramips/dts/mt7621_ruijie_rg-ew1200g-pro.dts:81]: The eeprom size should be `0x4da8` for mt7615.
```suggestion
						reg = <0x0 0x4da8>;
```

File: target/linux/ramips/dts/mt7621_ruijie_rg-ew1200g-pro.dts [hunk 5]
```diff
@@ -0,0 +1,142 @@
...
+				read-only;
+			};
+
+			partition@80000 {
+				label = "kdump";
+				reg = <0x80000 0x10000>;
+				read-only;
+			};
+
+			firmware: partition@90000 {
```
Comment (DragonBluep) [inline target/linux/ramips/dts/mt7621_ruijie_rg-ew1200g-pro.dts:98]:
```suggestion
			partition@90000 {
```

File: target/linux/ramips/dts/mt7621_ruijie_rg-ew1200g-pro.dts [hunk 6]
```diff
@@ -0,0 +1,142 @@
...
+		nvmem-cells = <&eeprom_factory_0>;
+		nvmem-cell-names = "eeprom";
+	};
+};
+
+&switch0 {
+	ports {
+		port@0 {
+			status = "okay";
+			label = "wan";
```
Comment (DragonBluep) [inline target/linux/ramips/dts/mt7621_ruijie_rg-ew1200g-pro.dts:124]: `wan` port can connect to gmac1. https://github.com/openwrt/openwrt/pull/10238

File: target/linux/ramips/image/mt7621.mk [hunk 7]
```diff
@@ -2109,6 +2109,16 @@ define Device/rostelecom_rt-sf-1
 endef
 TARGET_DEVICES += rostelecom_rt-sf-1
 
+define Device/ruijie_rg-ew1200g-pro
+  $(Device/dsa-migration)
+  $(Device/uimage-lzma-loader)
+  IMAGE_SIZE := 15808k
+  DEVICE_VENDOR := Ruijie
+  DEVICE_MODEL := RG-EW1200G PRO
+  DEVICE_PACKAGES := kmod-mt7615e kmod-mt7615-firmware
```
Comment (DragonBluep) [inline target/linux/ramips/image/mt7621.mk:2118]:
```suggestion
  DEVICE_PACKAGES := kmod-mt7615-firmware
```

File: target/linux/ramips/mt7621/base-files/etc/board.d/02_network [hunk 8]
```diff
@@ -159,6 +159,9 @@ ramips_setup_interfaces()
 		ucidef_set_interface_lan "lan1 lan2 lan3 lan4"
 		ucidef_set_interface "qtn" ifname "eth1" protocol "static" ipaddr "1.1.1.1" netmask "255.255.255.0"
 		;;
+	ruijie,rg-ew1200g-pro)
+		ucidef_set_interfaces_lan_wan "lan3 lan2 lan1" "wan"
```
Comment (DragonBluep) [inline target/linux/ramips/mt7621/base-files/etc/board.d/02_network:163]: merge it to case `ucidef_set_interfaces_lan_wan "lan1 lan2 lan3" "wan"`

File: target/linux/ramips/mt7621/base-files/etc/board.d/02_network [hunk 9]
```diff
@@ -303,6 +306,11 @@ ramips_setup_macs()
 		wan_mac=$label_mac
 		lan_mac=$(macaddr_add $label_mac 1)
 		;;
+	ruijie,rg-ew1200g-pro)
```
Comment (DragonBluep) [inline target/linux/ramips/mt7621/base-files/etc/board.d/02_network:309]: sort it alphabetically.

**Review by DragonBluep — COMMENTED** [review]
Please correct commit message. You can try:
```
git config --local user.name "YOUR NAME"
git config --local user.email "YOUR EMAIL"
git commit --amend -s --author="YOUR NAME <YOUR EMAIL>"
```

File: target/linux/ramips/mt7621/base-files/etc/board.d/02_network (hunk 8 above)
Comment (JakauBahuslau) [reply target/linux/ramips/mt7621/base-files/etc/board.d/02_network:163]: But ports on the router are labeled as "WAN LAN3 LAN2 LAN1" (if reading from left to right)

File: target/linux/ramips/dts/mt7621_ruijie_rg-ew1200g-pro.dts (hunk 2 above)
Comment (JakauBahuslau) [reply target/linux/ramips/dts/mt7621_ruijie_rg-ew1200g-pro.dts:52]: Applying that patch causes MTD partitions not to mount during boot with folowing error:

`spi0.0: error parsing ofpart partition /palmbus@1e000000/spi@b00/flash@0/partitions/partition@0 (/palmbus@1e000000/spi@b00/flash@0)`

Comment (DragonBluep) [reply target/linux/ramips/dts/mt7621_ruijie_rg-ew1200g-pro.dts:52]: These two lines need to be moved to `partitions` child node.

File: target/linux/ramips/mt7621/base-files/etc/board.d/02_network (hunk 8 above)
Comment (DragonBluep) [reply target/linux/ramips/mt7621/base-files/etc/board.d/02_network:163]: If the device has Ethernet port number labels, there is no need to do so. At least I didn't see any devices sorting port sequences in OpenWrt.

File: target/linux/ramips/dts/mt7621_ruijie_rg-ew1200g-pro.dts (hunk 2 above)
Comment (JakauBahuslau) [reply target/linux/ramips/dts/mt7621_ruijie_rg-ew1200g-pro.dts:52]: Now there is:

Comment (JakauBahuslau) [reply target/linux/ramips/dts/mt7621_ruijie_rg-ew1200g-pro.dts:52]: In fact, when I started porting, these lines were in `partitions` child note, and "bad cell count" error was fixed by moving these lines to top.

**Comment by musashino205** [comment]
Please drop the last 2x commits (merge/revert) and use ```git rebase``` if you want to update the branch.

**Pull request closed without merging**